kernel/cClosure.ml

```diff
@@ -1344,25 +1386,113 @@ let conv : (clos_infos -> clos_tab -> fconstr -> fconstr -> bool) ref
   = ref (fun _ _ _ _ -> (assert false : bool))
 let set_conv f = conv := f
 
+type 'constr partial_subst = {
+  subst: 'constr list;
+  usubst: Sorts.Quality.t list * Univ.Universe.t list;
+  rhs: constr;
+}
+
+type 'constr subst_status = Dead | Live of 'constr partial_subst
+
+type 'a status =
+  | Check of 'a
+  | Ignore
+
+module Status = struct
+  let map f = function Check a -> Check (f a) | Ignore as p -> p
+
+  let split = function Check (a, b) -> Check a, Check b | Ignore as p -> p, p
+  let split3 = function Check (a, b, c) -> Check a, Check b, Check c | Ignore as p -> p, p, p
+  let split4 = function Check (a, b, c, d) -> Check a, Check b, Check c, Check d | Ignore as p -> p, p, p, p
+  let split_array n = function
+  | Check a when Array.length a <> n -> invalid_arg "Status.split_array"
+  | Check a -> Array.init n (fun i -> Check (Array.unsafe_get a i))
+  | Ignore as p -> Array.make n p
+
+  let fold_left f a = function Check b -> f a b | Ignore -> a
+end
+
+type ('a, 'b) next =
+  | Continue of 'a
+  | Return of 'b
+
+type (_, _) escape =
+  | No:  ('i, 'i) escape
+  | Yes: ('i -> 'ret) * 'ret -> ('i, 'ret) escape
+
+type ('constr, 'stack, 'context) state =
+  | LocStart of { elims: pattern_elimination list status array; context: 'context; head: 'constr; stack: 'stack; next: ('constr, 'stack, 'context) state_next }
+  | LocArg of { patterns: pattern_argument status array; context: 'context; arg: 'constr; next: ('constr, 'stack, 'context) state }
+
+and ('constr, 'stack, 'context) state_next = (('constr, 'stack, 'context) state, bool * 'constr * 'stack) next
+
+
+type ('constr, 'stack, 'context) resume_state =
+  { states: 'constr subst_status array; context: 'context; patterns: head_elimination status array; next: ('constr, 'stack, 'context) state }
+
+type ('constr, 'stack, 'context, _) depth =
+  | Nil: ('constr * 'stack, 'ret) escape -> ('constr, 'stack, 'context, 'ret) depth
+  | Cons: ('constr, 'stack, 'context) resume_state * ('constr, 'stack, 'context, 'ret) depth -> ('constr, 'stack, 'context, 'ret) depth
+
+let extract_or_kill filter a status =
+  (* Array.split @@ Array.map2 step a status
+   * with step elim status -> match elim, status with
+   * | Ignore, s -> Ignore, s
+   * | Check e, s when filter e = None -> Ignore, Dead
+   * | Check e, s when filter e = Check p -> Check p', s
+   *)
```

Comment:
what does this mean?

Reply:
This is the operation that is done right below, but expressed in a more readable way. With how unoptimized the rest of my code is here, I should probably just use this code instead of the marginally faster one